src/filepath.c

```diff
@@ -2120,90 +2146,136 @@ f_resolve(typval_T *argvars, typval_T *rettv)
 	{
 	    for (;;)
 	    {
-		len = readlink((char *)p, (char *)buf, MAXPATHL);
-		if (len <= 0)
+		ssize_t rv = readlink((char *)p, (char *)buf, MAXPATHL);
+		if (rv <= 0)
 		    break;
-		buf[len] = NUL;
 
 		if (limit-- == 0)
 		{
 		    vim_free(p);
+		    vim_free(buf);
 		    vim_free(remain);
 		    emsg(_(e_too_many_symbolic_links_cycle));
 		    rettv->vval.v_string = NULL;
 		    goto fail;
 		}
 
+		buflen = (size_t)rv;
+		buf[buflen] = NUL;
+
 		// Ensure that the result will have a trailing path separator
 		// if the argument has one.
-		if (remain == NULL && has_trailing_pathsep)
-		    add_pathsep(buf);
+		if (remain == NULL && has_trailing_pathsep && !after_pathsep(buf, buf + buflen))
+		{
+		    STRCPY(buf + buflen, PATHSEPSTR);
+		    ++buflen;
+		}
 
 		// Separate the first path component in the link value and
 		// concatenate the remainders.
 		q = getnextcomp(vim_ispathsep(*buf) ? buf + 1 : buf);
 		if (*q != NUL)
 		{
+		    char_u  *q_prev = q - 1;
+
+		    // getnextcomp() finds the first path separator.
+		    // if there is a run of >1 path separators, set all
+		    // but the last in the run to NUL.
+		    while (*q != NUL && vim_ispathsep(*q))
+		    {
+			*q_prev = NUL;
+			q_prev = q;
+			MB_PTR_ADV(q);
+		    }
+		    q = q_prev;
+
 		    if (remain == NULL)
-			remain = vim_strsave(q - 1);
+		    {
+			rlen = (size_t)(buflen - (q - buf));
+			r = remain = vim_strnsave(q, rlen);
+			if (remain == NULL)
+			    rlen = 0;
+		    }
 		    else
 		    {
-			cpy = concat_str(q - 1, remain);
+			len = (size_t)(buflen - (q - buf));
+			cpysize = (size_t)(len + rlen + 1);		// +1 for NUL
 			if (cpy != NULL)
 			{
+			    rlen = (int)vim_snprintf((char *)cpy, cpysize, "%.*s%s", (int)len, q, r);
 			    vim_free(remain);
-			    remain = cpy;
+			    r = remain = cpy;
 			}
 		    }
-		    q[-1] = NUL;
+		    *q = NUL;
+		    buflen = (size_t)(q - buf);
 		}
 
 		q = gettail(p);
 		if (q > p && *q == NUL)
 		{
 		    // Ignore trailing path separator.
-		    p[q - p - 1] = NUL;
+		    plen = (size_t)(q - p - 1);
+		    p[plen] = NUL;
 		    q = gettail(p);
 		}
 		if (q > p && !mch_isFullName(buf))
 		{
 		    // symlink is relative to directory of argument
-		    cpy = alloc(STRLEN(p) + STRLEN(buf) + 1);
+		    cpy = alloc(plen + buflen + 1);
 		    if (cpy != NULL)
```

Comment:
do we need to handle the else case?

Reply:
The original code does not. However, we probably should. Taking our cue from other parts of the code, we can perform a `goto fail;`